File: java/99-project/qzone-crawler/trunk/src/main/java/exp/crawler/qq/core/_Mood.java
package exp.crawler.qq.core;

import java.io.File;

import net.sf.json.JSONArray;
import net.sf.json.JSONObject;

import org.openqa.selenium.WebDriver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import exp.crawler.qq.cache.Browser;
import exp.crawler.qq.utils.UIUtils;
import exp.libs.envm.Charset;
import exp.libs.utils.io.FileUtils;
import exp.libs.utils.other.StrUtils;
import exp.libs.utils.time.TimeUtils;
import exp.libs.utils.verify.RegexUtils;
import exp.libs.warp.net.http.HttpURLUtils;
import exp.libs.warp.net.webkit.WebBrowser;

public class _Mood {

	private final static Logger log = LoggerFactory.getLogger(MoodAnalyzer.class);
	
	private final static String INFO_DIR = "./data/page/";
	
	private final static String INFO_PATH_PREFIX = "page-info-";
	
	private final static String INFO_PATH_SUFFIX = ".txt";
	
	private final static String LINE_END = "\r\n";
	
	private final static String MSG_SPLIT = "##==========";
	
	private final static String IMAGE_DIR = "./data/img/";
	
	private final static String IMAGE_SUFFIX = ".jpg";
	
	private final static String IMAGE_ALL_DIR = "./data/img/all-pic/";
	
	/**
	 * 捕获在线数据, 存储图文信息到本地，作为离线数据备用
	 * @param loginPage
	 * @param username
	 * @param password
	 * @param targetQQ
	 * @param totalPageNum
	 */
	public static void catchOnlineInfo(WebBrowser bDriver, 
			String username, String password, 
			String targetQQ, int totalPageNum) {
		WebDriver driver = bDriver.getDriver();
		String gtk = Browser.GTK();
		
		if(StrUtils.isNotEmpty(gtk)) {
			UIUtils.log("登陆成功： 开始抓取 【说说】 数据...");
			FileUtils.delete(INFO_DIR);	// 删除旧数据
			
			for(int idx = 0; idx < totalPageNum; idx++) {
				UIUtils.log("正在抓取第 [", (idx + 1), "] 页的图文数据...");
				String pageUrl = getPageUrl(targetQQ, idx, gtk);
				driver.get(pageUrl);
				
				String pageData = driver.getPageSource();
				boolean isEnd = parse(idx, pageData);
				UIUtils.log("抓取第 [", (idx + 1), "] 页的图文数据完成.");
				
				if(isEnd) {
					break;
				}
			}
		} else {
			UIUtils.log("登陆失败： 无法获得 [g_tk] 码, 本次操作终止.");
		}
		
		bDriver.quit();
	}
	
	/**
	 * 获取【说说】的翻页地址
	 * @param qqNum
	 * @param pageNum
	 * @param token 似乎每次登陆都不同，但该次登陆之后就固定了
	 * @return
	 */
	protected static String getPageUrl(String qqNum, int pageNum, String token) {
		final int EACH_PAGE_LIMIT = 20;
		int startIdx = pageNum * EACH_PAGE_LIMIT;
		
		String url = StrUtils.concat("https://h5.qzone.qq.com/proxy/domain/", 
				"taotao.qq.com/cgi-bin/emotion_cgi_msglist_v6?uin=", qqNum,  
				"&inCharset=utf-8&outCharset=utf-8&hostUin=", qqNum,  
				"&notice=0&sort=0&pos=", startIdx, "&num=", EACH_PAGE_LIMIT,  
				"&cgi_host=http%3A%2F%2Ftaotao.qq.com%2Fcgi-bin", 
				"%2Femotion_cgi_msglist_v6&code_version=1&format=jsonp", 
				"&need_private_comment=1&g_tk=", token);
		return url;
	}
	
	/**
	 * 解析页面数据，存储到本地备用
	 * @param pageIdx
	 * @param pageData
	 */
	protected static boolean parse(int pageIdx, String pageData) {
		String filePath = getSaveFilePath(pageIdx);
		
		boolean isEnd = false;
		try {
			String json = getJson(pageData);
			JSONObject jsonObj = JSONObject.fromObject(json);
			JSONArray msglist = jsonObj.getJSONArray("msglist");
			for(int i = 0; i < msglist.size(); i++) {
				JSONObject msg = msglist.getJSONObject(i);
				save(msg, filePath);
			}
		} catch (Exception e) {
			
			// 已到了最末页
			if("JSONObject[\"msglist\"] is not a JSONArray.".equals(e.getMessage())) {
				isEnd = true;
				
			} else {
				log.error("抓取第 [{}] 页的图文数据失败.", pageIdx, e);
			}
		}
		return isEnd;
	}
	
	private static String getSaveFilePath(int pageIdx) {
		String filePath = StrUtils.concat(
				INFO_DIR, INFO_PATH_PREFIX, 
				StrUtils.leftPad(String.valueOf(pageIdx), '0', 3), 
				INFO_PATH_SUFFIX);
		return filePath;
	}
	
	/**
	 * 从html页面中截取真正有用的 json数据
	 * @param pageData
	 * @return
	 */
	private static String getJson(String pageData) {
		String regex = ">_Callback\\(([\\s\\S]+)\\);</pre></body></html>";
		String json = RegexUtils.findFirst(pageData, regex).replace("\\/", "/");
		return json;
	}
	
	private static void save(JSONObject msg, String filePath) {
		StringBuilder sb = new StringBuilder();
		
		String content = msg.getString("content").replaceAll("[\r|\n]", "").trim();
		sb.append(StrUtils.isEmpty(content) ? "null" : content).append(LINE_END);
		
		long createTime = msg.getLong("created_time") * 1000;
		sb.append(TimeUtils.toStr(createTime, "yyyyMMddHHmmss")).append(LINE_END);
		
		try {
			JSONArray pics = msg.getJSONArray("pic");
			for(int i = 0; i < pics.size(); i++) {
				JSONObject pic = pics.getJSONObject(i); 
				String url = pic.getString("url3");
				sb.append(url).append(LINE_END);
			}
		} catch (Exception e) {
			// Undo 不存在节点时会报错, 图片节点不一定存在，可忽略
		}
		
		sb.append(MSG_SPLIT).append(LINE_END);
		FileUtils.write(filePath, sb.toString(), Charset.UTF8, true);
	}
	
	/**
	 * 根据捕获的离线数据，下载图文信息并归整
	 */
	public static void downloadDatas() {
		int picNum = 0;
		UIUtils.log("准备下载图文信息, 正在删除历史数据...");
		FileUtils.delete(IMAGE_DIR);
		
		UIUtils.log("开始下载图文信息...");
		File pageInfoDir = new File(INFO_DIR);
		File[] pageInfos = pageInfoDir.listFiles();
		for(File pageInfo : pageInfos) {
			if(pageInfo.getName().startsWith(INFO_PATH_PREFIX) && 
					pageInfo.getName().endsWith(INFO_PATH_SUFFIX)) {
				picNum += down(pageInfo);
			}
		}
		UIUtils.log("下载所有图文信息完成, 共下载 [", picNum, "] 张图片.");
	}
	
	/**
	 * 解析并下载离线的页面信息文件
	 * @param pageInfo
	 */
	protected static int down(File pageInfo) {
		int picNum = 0;
		String fileName = pageInfo.getName().replace(INFO_PATH_SUFFIX, "");
		UIUtils.log("正在下载 [", fileName, "] 记录的图文信息...");
		
		String info = FileUtils.read(pageInfo, Charset.UTF8);
		String[] msgs = info.split(MSG_SPLIT);
		for(int idx = 0; idx < msgs.length; idx++) {
			String msg = msgs[idx].trim();
			if(StrUtils.isEmpty(msg)) {
				continue;
			}
			String msgPath = StrUtils.concat(fileName, "/", 
					StrUtils.leftPad(String.valueOf(idx), '0', 2));
			picNum += down(msgPath, msg);
		}
		UIUtils.log("下载 [", fileName, "] 记录的图文信息完成, 本页共 [", picNum, "] 张图片.");
		return picNum;
	}
	
	/**
	 * 下载单条【说说】信息
	 * @param msg
	 */
	private static int down(String filePath, String msg) {
		int picNum = 0;
		String saveDir = StrUtils.concat(IMAGE_DIR, filePath, "/");
		
		// 保存【说说】原文
		String msgPath = StrUtils.concat(saveDir, "msg.txt");
		FileUtils.write(msgPath, msg, Charset.UTF8, false);
		
		// 保存【说说】图片
		String[] lines = msg.split(LINE_END);
		String createTime = lines[1];
		
		UIUtils.log("正在下载说说 [", lines[0], "] 记录的图文信息...");
		if(lines.length > 2) {
			for(int i = 2; i < lines.length; i++) {
				String line = lines[i].trim();
				if(!line.startsWith("http")) {
					continue;
				}
				
				String picName = StrUtils.concat(createTime, "-", i - 2, IMAGE_SUFFIX);
				String savePath = StrUtils.concat(saveDir, picName);
				HttpURLUtils.downloadByGet(savePath, line, null, null);
				picNum++;
			}
		}
		UIUtils.log("下载图文信息完成, 本条【说说】共 [", picNum, "] 张图片.");
		return picNum;
	}
	
	/**
	 * 把所有下载的图片复制到同一个目录
	 */
	public static void copyTogether() {
		FileUtils.delete(IMAGE_ALL_DIR);
		FileUtils.createDir(IMAGE_ALL_DIR);
		
		UIUtils.log("正在复制 [", IMAGE_DIR, "] 下的所有图片到 [", IMAGE_ALL_DIR, "] ...");
		copy(new File(IMAGE_DIR));
		UIUtils.log("复制完成.");
	}
	
	private static void copy(File file) {
		if(file.isFile()) {
			if(file.getName().endsWith(IMAGE_SUFFIX)) {
				FileUtils.copyFile(file, 
						new File(StrUtils.concat(IMAGE_ALL_DIR, file.getName())));
			}
		} else {
			File[] files = file.listFiles();
			for(File f : files) {
				copy(f);
			}
		}
	}
}

File: python/99-project/qzone-crawler/src/main/py/core/mood.py
# -*- coding: utf-8 -*-
__author__ = 'EXP (272629724@qq.com)'
__date__ = '2018-03-29 20:17'

import os
import shutil
import requests
import time
import json
import traceback
import src.main.py.config as cfg
import src.main.py.utils.xhr as xhr
import src.main.py.utils.pic as pic
from src.main.py.bean.cookie import QQCookie


class MoodAnalyzer(object):
    '''
    【空间说说】解析器
    '''

    MOOD_INFO_NAME = 'MoodInfo-[说说信息].txt'  # 说说分页信息保存文件名
    cookie = None           # 已登陆的QQCookie
    QQ = ''                 # 被爬取数据的目标QQ
    MOOD_DIR = ''           # 说说保存目录
    PAGE_DIR_PREFIX = ''    # 说说每页图文信息的保存路径前缀
    PHOTO_DIR = ''          # 说说所有照片的保存目录


    def __init__(self, cookie, QQ):
        '''
        构造函数
        :param cookie: 已登陆的QQCookie
        :param QQ: 被爬取数据的目标QQ
        :return:None
        '''
        self.cookie = QQCookie() if not cookie else cookie
        self.QQ = '0' if not QQ else QQ.strip()
        self.MOOD_DIR = '%s%s/mood/' % (cfg.DATA_DIR, self.QQ)
        self.PAGE_DIR_PREFIX = '%scontent/page-' % self.MOOD_DIR
        self.PHOTO_DIR = '%sphotos/' % self.MOOD_DIR


    def execute(self):
        '''
        执行空间说说解析, 并下载所有说说及相关照片
        :return:
        '''
        try:

            # 清除上次下载的数据
            if os.path.exists(self.MOOD_DIR):
                shutil.rmtree(self.MOOD_DIR)
            os.makedirs(self.MOOD_DIR)

            # 下载说说及照片
            moods = self.get_moods()


            print('任务完成: QQ [%s] 的空间说说已保存到 [%s]' % (self.QQ, self.MOOD_DIR))

        except:
            print('任务失败: 下载 QQ [%s] 的空间说说时发生异常' % self.QQ)
            traceback.print_exc()


    def get_moods(self):
        '''
        提取所有说说及相关的照片信息
        :return: 说说列表（含相关照片信息）
        '''
        print('正在提取QQ [%s] 的说说动态...' % self.QQ)
        moods = []
        PAGE_NUM = self.get_page_num()
        for page in range(PAGE_NUM) :
            page += 1
            print(' -> 正在提取第 [%d/%d] 页的说说信息...' % (page, PAGE_NUM))

            # TODO

            print(' -> 第 [%d/%d] 页说说提取完成, 累计说说数量: %d' % (page, PAGE_NUM, len(moods)))
            time.sleep(cfg.SLEEP_TIME)

        return moods


    def get_page_num(self):
        '''
        获取说说总页数
        :return: 说说总页数
        '''
        print('正在提取QQ [%s] 的说说页数...' % self.QQ)
        total = 0
        try:
            root = json.loads(self.get_page_moods_json(1))
            total = root.get('total', 0)

        except:
            print('提取QQ [%s] 的说说页数失败' % self.QQ)
            traceback.print_exc()

        return total


    def get_page_moods(self, page):
        '''
        获取分页的说说内容
        :param page: 页码
        :return: 分页说说列表（含相关照片信息）
        '''
        moods = []
        try:
            root = json.loads(self.get_page_moods_json(page))
            msg_list = root['msglist']
            for msg in msg_list :
                content = msg.get('content', '')
                create_time = msg.get('created_time', 0) * 1000


        except:
            print('提取第 [%d] 页的说说信息异常' % page)
            traceback.print_exc()

        return moods

# 			JSONObject json = JSONObject.fromObject(response);
# 			JSONArray msglist = JsonUtils.getArray(json, XHRAtrbt.msglist);
# 			for(int i = 0; i < msglist.size(); i++) {
# 				JSONObject msg = msglist.getJSONObject(i);
# 				String content = JsonUtils.getStr(msg, XHRAtrbt.content);
# 				long createTime = JsonUtils.getLong(msg, XHRAtrbt.created_time, 0) * 1000;
#
# 				Mood mood = new Mood(page, content, createTime);
# 				JSONArray pics = JsonUtils.getArray(msg, XHRAtrbt.pic);
# 				for(int j = 0; j < pics.size(); j++) {
# 					JSONObject pic = pics.getJSONObject(j);
# 					String url = JsonUtils.getStr(pic, XHRAtrbt.url3);
# 					url = PicUtils.convert(url);
# 					mood.addPicURL(url);
# 				}
# 				moods.add(mood);
# 	}


    def get_page_moods_json(self, page):
        '''
        获取分页的说说Json
        :param page: 页码
        :return: 分页的说说Json
        '''
        headers = xhr.get_headers(self.cookie.to_nv())
        headers['Referer'] = cfg.MOOD_REFERER
        params = {
            'g_tk' : self.cookie.gtk,
            'qzonetoken' : self.cookie.qzone_token,
            'uin' : self.QQ,
            'hostUin' : self.QQ,
            'pos' : '%d' % ((page - 1) * cfg.BATCH_LIMT),
            'num' : '%d' % cfg.BATCH_LIMT,
            'cgi_host' : cfg.MOOD_DOMAIN,
            'inCharset' : cfg.DEFAULT_CHARSET,
            'outCharset' : cfg.DEFAULT_CHARSET,
            'notice' : '0',
            'sort' : '0',
            'code_version' : '1',
            'format' : 'jsonp',
            'need_private_comment' : '1',
        }
        response = requests.get(cfg.MOOD_URL, headers=headers, params=params)
        return xhr.to_json(response.text)


#
# 	/**
# 	 * 下载所有说说及相关的照片
# 	 * @param moods 说说集（含照片信息）
# 	 */
# 	private void download(List<Mood> moods) {
# 		if(ListUtils.isEmpty(moods)) {
# 			return;
# 		}
#
# 		UIUtils.log("提取QQ [", QQ, "] 的说说及照片完成, 开始下载...");
# 		int idx = 1;
# 		for(Mood mood : moods) {
# 			UIUtils.log("正在下载第 [", idx++, "/", moods.size(), "] 条说说: ", mood.CONTENT());
# 			int cnt = _download(mood);
# 			boolean isOk = (cnt == mood.PIC_NUM());
# 			UIUtils.log(" -> 说说照片下载完成, 成功率: ", cnt, "/", mood.PIC_NUM());
#
# 			// 保存下载信息
# 			String savePath = StrUtils.concat(PAGE_DIR_PREFIX, mood.PAGE(), "/", MOOD_INFO_NAME);
# 			FileUtils.write(savePath, mood.toString(isOk), Config.CHARSET, true);
# 		}
# 	}
#
# 	/**
# 	 * 下载单条说说及相关的照片
# 	 * @param mood 说说信息
# 	 * @return 成功下载的照片数
# 	 */
# 	private int _download(Mood mood) {
# 		Map<String, String> header = XHRUtils.getHeader(Browser.COOKIE());
# 		header.put(HttpUtils.HEAD.KEY.REFERER, URL.MOOD_REFERER);
#
# 		int idx = 0, cnt = 0;
# 		for(String picURL : mood.getPicURLs()) {
# 			String picName = PicUtils.getPicName(String.valueOf(idx++), mood.CONTENT());
# 			boolean isOk = _download(header, mood.PAGE(), picName, picURL);
# 			cnt += (isOk ? 1 : 0);
#
# 			UIUtils.log(" -> 下载照片进度(", (isOk ? "成功" : "失败"), "): ", cnt, "/", mood.PIC_NUM());
# 		}
# 		return cnt;
# 	}
#
# 	/**
# 	 * 下载单张图片到说说的分页目录，并复制到图片合集目录
# 	 * @param header
# 	 * @param pageIdx 页码索引
# 	 * @param picName
# 	 * @param picURL
# 	 * @return
# 	 */
# 	private boolean _download(Map<String, String> header,
# 			String pageIdx, String picName, String picURL) {
# 		header.put(HttpUtils.HEAD.KEY.HOST, XHRUtils.toHost(picURL));
#
# 		boolean isOk = false;
# 		String savePath = StrUtils.concat(PAGE_DIR_PREFIX, pageIdx, "/", picName);
# 		for(int retry = 0; !isOk && retry < Config.RETRY; retry++) {
# 			isOk = HttpURLUtils.downloadByGet(savePath, picURL, header, null,
# 					Config.TIMEOUT, Config.TIMEOUT, Config.CHARSET);
#
# 			if(isOk == false) {
# 				FileUtils.delete(savePath);
# 				ThreadUtils.tSleep(Config.SLEEP_TIME);
#
# 			} else {
# 				FileUtils.copyFile(savePath, PHOTO_DIR.concat(picName));
# 			}
# 		}
# 		return isOk;
# 	}
#